packages/hash/frontend/src/blocks/page/Comments/CommentButton.tsx

```diff
@@ -0,0 +1,88 @@
+import { FunctionComponent, useCallback, useRef, useState } from "react";
+import { IconButton, FontAwesomeIcon } from "@hashintel/hash-design-system";
+import { faComment } from "@fortawesome/free-regular-svg-icons";
+import Box from "@mui/material/Box";
+import { TextToken } from "@hashintel/hash-shared/graphql/types";
+import Popper from "@mui/material/Popper";
+import { useBlockView } from "../BlockViewContext";
+import { useCreateComment } from "../../../components/hooks/useCreateComment";
+import { useRouteAccountInfo } from "../../../shared/routing";
+import { CommentTextField } from "./CommentTextField";
+
+type CommentButtonProps = {
+  blockId: string | null;
+  className: string;
+  rootNode: HTMLElement;
+};
+
+export const CommentButton: FunctionComponent<CommentButtonProps> = ({
+  blockId,
+  className,
+  rootNode,
+}) => {
+  const { accountId } = useRouteAccountInfo();
+  const blockView = useBlockView();
+  const [open, setOpen] = useState(false);
+  const anchorRef = useRef<HTMLButtonElement>(null);
+  const [createComment] = useCreateComment(accountId);
+
+  const submitComment = useCallback(
+    async (content: TextToken[]) => {
+      if (blockId) {
+        await createComment(blockId, content);
+      }
+    },
+    [createComment, blockId],
+  );
+
+  const closeInput = () => setOpen(false);
+
+  const anchorNode = anchorRef.current;
```

Comment:
Reading from refs during render isn't safe, because their value isn't updated by React until after render is called and updating them doesn't trigger another render. It's also not concurrent safe. 

You use this in two places – in calculating the offset, you can just read from the ref at that point in time. 

If you need the node to pass as the anchor el to the popper, you should use state to store the node, not a ref. 

See the example from the MUI docs

```ts
import * as React from 'react';
import Box from '@mui/material/Box';
import Popper from '@mui/material/Popper';

export default function SimplePopper() {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(anchorEl ? null : event.currentTarget);
  };

  const open = Boolean(anchorEl);
  const id = open ? 'simple-popper' : undefined;

  return (
    <div>
      <button aria-describedby={id} type="button" onClick={handleClick}>
        Toggle Popper
      </button>
      <Popper id={id} open={open} anchorEl={anchorEl}>
        <Box sx={{ border: 1, p: 1, bgcolor: 'background.paper' }}>
          The content of the Popper.
        </Box>
      </Popper>
    </div>
  );
}
```

Reply:
Thank you, changed it to use state instead of a ref in 33cf4d52941c54919c4c9ac9036e9b9f897fa55a.